src/rulesets/Base/indexing.jl

```diff
@@ -26,3 +26,43 @@ function rrule(::typeof(getindex), x::Array{<:Number}, inds...)
 
     return y, getindex_pullback
 end
+
+
+#####
+##### `eachslice` and friends
+#####
+
+function rrule(::typeof(eachrow), x::AbstractVecOrMat)
+    allrows(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, 1))
+    return collect(eachrow(x)), allrows
+end
+
+function rrule(::typeof(eachcol), x::AbstractVecOrMat)
+    allcols(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, 2))
+    return collect(eachcol(x)), allcols
+end
+
+function rrule(::typeof(eachslice), x::AbstractArray; dims::Integer)
+    allslices(dy) = (NoTangent(), ∇eachslice(unthunk(dy), x, dims))
+    return collect(eachslice(x; dims=dims)), allslices
+end
+
+function ∇eachslice(dys, x::AbstractArray, dim::Integer)
+    i1 = findfirst(dy -> dy isa AbstractArray, dys)
+    if i1 === nothing  # all slices are Zero!
+        return (zero(x),)
+    end
+    T = promote_type(eltype(dys[i1]), eltype(x))
+    # The whole point of this gradient is that we can allocate one `dx` array:
+    dx = similar(x, T)
+    for i in axes(x, dim)
+        slice = selectdim(dx, dim, i)
```

Comment:
What I meant to do besides fixing tests is checking whether this is fast. `selectdim` isn't type-stable, maybe constant prop. will save us, maybe not. `eachcol` could pass a `Val(2)` etc.

Reply:
Worth doing, about a factor of 2. Not completely type-stable.